File: backend/app/kpis_email.py
"""KPIs ActiveCampaign — calculés depuis la table ac_campaigns dans DuckDB."""
from datetime import date, timedelta

from . import cache
from .kpis import _get_target, _status


def _safe(val, default=None):
    if val is None:
        return default
    if hasattr(val, "item"):
        return val.item()
    return val


def _table_exists(name: str) -> bool:
    try:
        cache.query_one(f"SELECT COUNT(*) FROM {name}")
        return True
    except Exception:
        return False


# ─── KPI Cards ───────────────────────────────────────────────────────────────

def open_rate(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    row = cache.query(f"""
        SELECT
            SUM(uniqueopens)::float / NULLIF(SUM(send_amt), 0) AS open_rate,
            COUNT(*) AS nb_campaigns
        FROM ac_campaigns
        WHERE sdate >= '{cutoff}'
    """)
    val = _safe(row.iloc[0]["open_rate"]) if not row.empty else None
    t = _get_target("open_rate_email")
    st = _status(val, t["target"] if t else None, t["seuil"] if t else None, t["sens"] if t else None)
    return {"value": val, "status": st}


def ctor(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    row = cache.query(f"""
        SELECT SUM(uniquelinkclicks)::float / NULLIF(SUM(uniqueopens), 0) AS ctor
        FROM ac_campaigns
        WHERE sdate >= '{cutoff}'
    """)
    val = _safe(row.iloc[0]["ctor"]) if not row.empty else None
    return {
        "value": val,
        "status": "green" if val and val >= 0.10 else ("orange" if val and val >= 0.05 else "red"),
    }


def unsubscribes(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    val = cache.query_one(f"""
        SELECT SUM(unsubscribes) FROM ac_campaigns WHERE sdate >= '{cutoff}'
    """)
    val = _safe(val, 0)
    return {
        "value": val,
        "status": "green" if val is not None and val < 100 else ("orange" if val is not None and val < 300 else "red"),
    }


def unsubscribe_rate(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    row = cache.query(f"""
        SELECT SUM(unsubscribes)::float / NULLIF(SUM(send_amt), 0) AS rate
        FROM ac_campaigns WHERE sdate >= '{cutoff}'
    """)
    val = _safe(row.iloc[0]["rate"]) if not row.empty else None
    return {
        "value": val,
        "status": "green" if val is not None and val < 0.002 else ("orange" if val is not None and val < 0.005 else "red"),
    }


def total_sends(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    val = cache.query_one(f"SELECT SUM(send_amt) FROM ac_campaigns WHERE sdate >= '{cutoff}'")
    return {"value": _safe(val, 0), "status": "unknown"}


def nb_campaigns(days: int = 30) -> dict:
    if not _table_exists("ac_campaigns"):
        return {"value": None, "status": "unknown"}
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    val = cache.query_one(f"SELECT COUNT(*) FROM ac_campaigns WHERE sdate >= '{cutoff}'")
    return {"value": _safe(val, 0), "status": "unknown"}


# ─── Charts ──────────────────────────────────────────────────────────────────

def chart_open_rate(days: int = 90) -> list:
    """OR par campagne sur les `days` derniers jours, trié par date."""
    if not _table_exists("ac_campaigns"):
        return []
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    df = cache.query(f"""
        SELECT
            sdate,
            name,
            send_amt,
            uniqueopens,
            uniquelinkclicks,
            open_rate,
            ctr,
            ctor,
            unsubscribes
        FROM ac_campaigns
        WHERE sdate >= '{cutoff}'
        ORDER BY sdate ASC
    """)
    if df.empty:
        return []
    return [
        {
            "date": str(row["sdate"]),
            "name": row["name"],
            "open_rate": round(float(row["open_rate"]), 4) if row["open_rate"] else 0,
            "ctr": round(float(row["ctr"]), 4) if row["ctr"] else 0,
        }
        for _, row in df.iterrows()
    ]


# ─── Tables ──────────────────────────────────────────────────────────────────

def campaigns_table(days: int = 90, limit: int = 50) -> list:
    """Top campagnes triées par date desc."""
    if not _table_exists("ac_campaigns"):
        return []
    cutoff = (date.today() - timedelta(days=days)).isoformat()
    df = cache.query(f"""
        SELECT
            name,
            sdate,
            send_amt,
            uniqueopens,
            uniquelinkclicks,
            unsubscribes,
            ROUND(open_rate * 100, 1) AS open_rate_pct,
            ROUND(ctr * 100, 1)       AS ctr_pct,
            ROUND(ctor * 100, 1)      AS ctor_pct
        FROM ac_campaigns
        WHERE sdate >= '{cutoff}'
        ORDER BY sdate DESC
        LIMIT {limit}
    """)
    if df.empty:
        return []
    return df.to_dict(orient="records")

File: workers/src/lib/kpis_email.ts
/**
 * KPIs email (ActiveCampaign) — port de backend/app/kpis_email.py.
 *
 * NOTE: la table `ac_campaigns` est vide tant que le pipeline d'ingestion
 * ActiveCampaign n'est pas en place côté Workers (Cron Trigger à venir).
 * Les fonctions retournent donc `{value: null, status: 'unknown'}` mais
 * ne plantent pas.
 */
import type postgres from 'postgres'
import type { StatusValue } from './kpi_helpers'

interface SimpleKpi {
  value: number | null
  status: StatusValue
}

function cutoffDate(days: number, today = new Date()): string {
  const t = new Date(Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate()))
  const cutoff = new Date(t.getTime() - days * 86_400_000)
  return cutoff.toISOString().slice(0, 10)
}

function asNumberOrNull(v: unknown): number | null {
  if (v === null || v === undefined) return null
  const n = typeof v === 'number' ? v : Number(v)
  return Number.isFinite(n) ? n : null
}

export async function openRate(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ open_rate: number | null }>>`
    SELECT SUM(uniqueopens)::float / NULLIF(SUM(send_amt), 0) AS open_rate
    FROM ac_campaigns WHERE sdate >= ${c}
  `
  const val = asNumberOrNull(rows[0]?.open_rate)
  let status: StatusValue = 'unknown'
  if (val !== null) {
    if (val >= 0.25) status = 'green'
    else if (val >= 0.15) status = 'orange'
    else status = 'red'
  }
  return { value: val, status }
}

export async function ctor(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ ctor: number | null }>>`
    SELECT SUM(uniquelinkclicks)::float / NULLIF(SUM(uniqueopens), 0) AS ctor
    FROM ac_campaigns WHERE sdate >= ${c}
  `
  const val = asNumberOrNull(rows[0]?.ctor)
  let status: StatusValue = 'unknown'
  if (val !== null) {
    if (val >= 0.10) status = 'green'
    else if (val >= 0.05) status = 'orange'
    else status = 'red'
  }
  return { value: val, status }
}

export async function unsubscribes(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ v: number | null }>>`
    SELECT COALESCE(SUM(unsubscribes),0) AS v FROM ac_campaigns WHERE sdate >= ${c}
  `
  const val = asNumberOrNull(rows[0]?.v)
  let status: StatusValue = 'unknown'
  if (val !== null) {
    if (val < 100) status = 'green'
    else if (val < 300) status = 'orange'
    else status = 'red'
  }
  return { value: val, status }
}

export async function unsubscribeRate(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ rate: number | null }>>`
    SELECT SUM(unsubscribes)::float / NULLIF(SUM(send_amt), 0) AS rate
    FROM ac_campaigns WHERE sdate >= ${c}
  `
  const val = asNumberOrNull(rows[0]?.rate)
  let status: StatusValue = 'unknown'
  if (val !== null) {
    if (val < 0.002) status = 'green'
    else if (val < 0.005) status = 'orange'
    else status = 'red'
  }
  return { value: val, status }
}

export async function totalSends(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ v: number | null }>>`
    SELECT COALESCE(SUM(send_amt),0) AS v FROM ac_campaigns WHERE sdate >= ${c}
  `
  return { value: asNumberOrNull(rows[0]?.v), status: 'unknown' }
}

export async function nbCampaigns(sql: postgres.Sql, days = 30): Promise<SimpleKpi> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ v: number | null }>>`
    SELECT COUNT(*) AS v FROM ac_campaigns WHERE sdate >= ${c}
  `
  return { value: asNumberOrNull(rows[0]?.v), status: 'unknown' }
}

export async function chartOpenRate(
  sql: postgres.Sql,
  days = 90,
): Promise<Array<{ date: string; name: string; open_rate: number; ctr: number }>> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ sdate: Date | string | null; name: string | null; open_rate: number | null; ctr: number | null }>>`
    SELECT sdate, name, open_rate, ctr FROM ac_campaigns WHERE sdate >= ${c} ORDER BY sdate
  `
  return rows
    .filter((r) => r.sdate !== null)
    .map((r) => ({
      date: (typeof r.sdate === 'string' ? r.sdate : (r.sdate as Date).toISOString()).slice(0, 10),
      name: r.name ?? '',
      open_rate: asNumberOrNull(r.open_rate) ?? 0,
      ctr: asNumberOrNull(r.ctr) ?? 0,
    }))
}

export async function campaignsTable(
  sql: postgres.Sql,
  days = 90,
  limit = 50,
): Promise<Array<{ name: string; sdate: string; send_amt: number; uniqueopens: number; uniquelinkclicks: number; unsubscribes: number; open_rate_pct: number; ctr_pct: number; ctor_pct: number }>> {
  const c = cutoffDate(days)
  const rows = await sql<Array<{ name: string | null; sdate: Date | string | null; send_amt: number | null; uniqueopens: number | null; uniquelinkclicks: number | null; unsubscribes: number | null }>>`
    SELECT name, sdate, send_amt, uniqueopens, uniquelinkclicks, unsubscribes
    FROM ac_campaigns WHERE sdate >= ${c}
    ORDER BY sdate DESC LIMIT ${limit}
  `
  return rows.map((r) => {
    const send = asNumberOrNull(r.send_amt) ?? 0
    const opens = asNumberOrNull(r.uniqueopens) ?? 0
    const clicks = asNumberOrNull(r.uniquelinkclicks) ?? 0
    const unsubs = asNumberOrNull(r.unsubscribes) ?? 0
    return {
      name: r.name ?? '',
      sdate: r.sdate ? (typeof r.sdate === 'string' ? r.sdate : r.sdate.toISOString()).slice(0, 10) : '',
      send_amt: send,
      uniqueopens: opens,
      uniquelinkclicks: clicks,
      unsubscribes: unsubs,
      open_rate_pct: send ? Math.round((opens / send) * 1000) / 10 : 0,
      ctr_pct: send ? Math.round((clicks / send) * 1000) / 10 : 0,
      ctor_pct: opens ? Math.round((clicks / opens) * 1000) / 10 : 0,
    }
  })
}
